src/Native/Unix/System.Native/pal_random.c

```diff
@@ -2,109 +2,269 @@
 #include <sys/stat.h>
-#include <fcntl.h>
-#include <assert.h>
-#include <unistd.h>
 #include <time.h>
-#include <errno.h>
+#include <unistd.h>
 
 #include "pal_config.h"
 #include "pal_random.h"
+
+#if defined(__linux__)
+#include <sys/syscall.h>
+#if defined(SYS_getrandom)
+# define PAL_GRND_NONBLOCK 1
+#endif // SYS_getrandom
+#endif // __linux__
+
 /*
 
 Generate random bytes. The generated bytes are not cryptographically strong.
 
 */
 
-void SystemNative_GetNonCryptographicallySecureRandomBytes(uint8_t* buffer, int32_t bufferLength)
+#if !HAVE_ARC4RANDOM_BUF
+static int GetUrandomFileDescriptor(void)
 {
-    assert(buffer != NULL);
-
-#if HAVE_ARC4RANDOM_BUF
-    arc4random_buf(buffer, (size_t)bufferLength);
-#else
-    static volatile int rand_des = -1;
-    long num = 0;
-    static bool sMissingDevURandom;
-    static bool sInitializedMRand;
+    static int fd = -1;
+    int newFd;
 
-    if (!sMissingDevURandom)
+    if (fd >= 0)
     {
-        if (rand_des == -1)
-        {
-            int fd;
+        return fd;
+    }
 
-            do
-            {
+    do
+    {
 #if HAVE_O_CLOEXEC
-                fd = open("/dev/urandom", O_RDONLY, O_CLOEXEC);
+        newFd = open("/dev/urandom", O_RDONLY | O_CLOEXEC);
 #else
-                fd = open("/dev/urandom", O_RDONLY);
-                fcntl(fd, F_SETFD, FD_CLOEXEC);
+        newFd = open("/dev/urandom", O_RDONLY);
+        fcntl(newFd, F_SETFD, FD_CLOEXEC);
 #endif
-            }
-            while ((fd == -1) && (errno == EINTR));
+    }
+    while ((newFd == -1) && (errno == EINTR));
 
-            if (fd != -1)
+    int unsetFd = -1;
+    if (!__atomic_compare_exchange_n(&fd, &unsetFd, newFd, false, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST))
+    {
+        close(newFd);
+    }
+
+    return fd;
+}
+
+static void XorWithLrand48(uint8_t *buffer, int32_t bufferLength)
+{
+    static bool sInitializedLrand;
+    size_t length = (size_t)bufferLength;
+
+    if (!sInitializedLrand)
+    {
+        bool lRandUninitialized = false;
+
+        if (__atomic_compare_exchange_n(&sInitializedLrand, &lRandUninitialized, true, false,
+                                        __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST))
+        {
+            // TODO(38034): There's not enough entropy in time(NULL) to seed the
+            // PRNG.
+            srand48(time(NULL));
+        }
+        else
+        {
+            // Setting the flag and initializing srand should be quick, so
+            // if we lose the race, it's very unlikely this will spin for
+            // too much time.
+            while (!__atomic_load_n(&sInitializedLrand, __ATOMIC_SEQ_CST))
             {
-                int expected = -1;
-                if (!__atomic_compare_exchange_n(&rand_des, &expected, fd, false, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST))
-                {
-                    // Another thread has already set the rand_des
-                    close(fd);
-                }
+                sched_yield();
```

Comment:
Is it possible that a normal priority thread won the compare exchange, but a high priority thread gets stuck in here?  Will that just deadlock, or will the high priority thread eventually yield anyways? (the description of sched_yield suggests that a high priority thread just ignores the yield call).